tests/test_client.py

```diff
@@ -0,0 +1,315 @@
+    def count_all(self):
+        return len(self.get_all())
+
+    def get_first(self):
+        return self.get_all()[0]
+
+    @abstractmethod
+    def get_all(self):
+        pass
+
+
+class TestProjects(TestStream):
+    tap_stream_id = 'projects'
+
+    def create_test_data(self, min_ensure_exists=1):
+        raise NotImplementedError("Not implemented yet")
+
+    def get_all(self):
+        return self._client.request(
+            self.tap_stream_id, "GET", "/rest/api/2/project",
+            params={"expand": "description,lead,url,projectKeys"})
+
+class TestUsers(TestStream):
+    tap_stream_id = 'users'
+
+    def create_test_data(self, min_ensure_exists=1):
+        raise NotImplementedError("Not implemented yet")
+
+    def get_all(self):
+        max_results = 2
+        groups = [
+            "jira-administrators",
+            "jira-software-users",
+        ]
+        all_users = list()
+        for group in groups:
+            params = {"groupname": group,
+                        "maxResults": max_results,
+                        "includeInactiveUsers": True}
+            pager = Paginator(self._client, items_key='values')
+            for page in pager.pages(self.tap_stream_id, "GET",
+                                    "/rest/api/2/group/member",
+                                    params=params):
+                for user in page:
+                    all_users.append(user)
+
+        return all_users
+
+class TestComponents(TestStream):
+    tap_stream_id = 'components'
+
+    def create_test_data(self, min_ensure_exists=1):
+        count = self.count_all()
+        LOGGER.info("Found %s records for stream %s", count, self.tap_stream_id)
+
+        if min_ensure_exists - count + 1 > 0:
+            LOGGER.info("Need to create %s more records for stream %s, doing so now", min_ensure_exists - count + 1, self.tap_stream_id)
+            for _ in range(0, min_ensure_exists - count + 1):
+                random_uuid = uuid.uuid4()
+                self._client.request(
+                    self.tap_stream_id, "POST", "/rest/api/2/component",
+                    headers={"Content-Type": "application/json"},
+                    data=json.dumps({
+                        "isAssigneeTypeValid": False,
+                        "name": "Component {}".format(random_uuid),
+                        "description": "This is Jira component {}".format(random_uuid),
+                        "project": self.get_test_project()['key'],
+                        "leadAccountId": self.get_test_user()['accountId'],
+                    }),
+                )
+
+    # Cache means it saves the result of this call forever
+    # It assumes project data does not change
+    @functools.lru_cache
+    def get_test_project(self):
+        return TestProjects(self._client).get_first()
+
+    # Cache means it saves the result of this call forever
+    # It assumes project data does not change
+    @functools.lru_cache
+    def get_test_user(self):
```

Comment:
Could this be defined either inside the function above or as a `__` method? To discourage outside use where the cache becomes an issue?

Reply:
Definitely agree with making the methods private. There's a reason I put it inside a method like this (intended only to be used by this instance) and not around the call to retrieve all the projects or users.

Reply:
Done in latest commit